client/state/routing/middleware.js

```diff
@@ -9,11 +9,18 @@ import page from 'page';
  */
 import localforage from 'lib/localforage';
 import { isOutsideCalypso } from 'lib/url';
-import { ROUTE_SET } from 'state/action-types';
+import { ROUTE_SET, ROUTE_TO } from 'state/action-types';
 
 const debug = debugFactory( 'calypso:restore-last-location' );
 const LAST_PATH = 'last_path';
 
+export const routingMiddleware = () => ( next ) => ( action ) => {
+	if ( action.type === ROUTE_TO ) {
+		page( action.path );
+	}
```

Comment:
@dmsnell I feel like this can be easily be split up into two PRs, one that adds the routing middleware so we can avoid direct page calls, and another that adds comment specific actions that we might happen to update the url as a side effect.

I think I'll drop e709e52 and 8b28e25 so the intent in this PR is a bit more clear.

Currently the latter PR feels very optional to me since the current filter is provided by our controller, though we can easily do something like what you suggest. The only downside being that we'd likely have to move some of the path generation into a util so it can be shared by the comment NavItem (which still needs a real link so right click works) and the middleware handler.